### PR TITLE
[AArch64] Consider histcnt smaller than i32 in the cost model

### PR DESCRIPTION
**Pull request by SamTebbs33**
This PR updates the AArch64 cost model to consider the cheaper cost of <i32 histograms to reflect the improvements from
https://github.com/llvm/llvm-project/pull/101017 and https://github.com/llvm/llvm-project/pull/103037

Work by Max Beck-Jones (@DevM-uk)

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 1]
```diff
@@ -517,25 +517,31 @@ static bool isUnpackedVectorVT(EVT VecVT) {
 static InstructionCost getHistogramCost(const IntrinsicCostAttributes &ICA) {
   Type *BucketPtrsTy = ICA.getArgTypes()[0]; // Type of vector of pointers
   Type *EltTy = ICA.getArgTypes()[1];        // Type of bucket elements
+  unsigned TotalHistCnts = 1;
 
-  // Only allow (32b and 64b) integers or pointers for now...
+  // Only allow (up to 64b) integers or pointers
```
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:522]: I'd suggest adding a variable `unsigned EltSize = EltTy->getScalarSizeInBits();` and using it both in the initial size check and the total size check below.

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 2]
```diff
@@ -517,25 +517,31 @@ static bool isUnpackedVectorVT(EVT VecVT) {
...
-         VTy->getElementCount().getKnownMinValue() != 4) ||
-        VTy->getPrimitiveSizeInBits().getKnownMinValue() > 128 ||
-        !VTy->isScalableTy())
+  if (VectorType *VTy = dyn_cast<VectorType>(BucketPtrsTy)) {
+    unsigned EC = VTy->getElementCount().getKnownMinValue();
+    if (!isPowerOf2_64(EC) || !VTy->isScalableTy())
       return InstructionCost::getInvalid();
 
-  return InstructionCost(BaseHistCntCost);
+    bool Element64b = EltTy->isIntegerTy(64);
```
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:534]: Using the `EltSize` suggestion from above:

```suggestion
    // HistCnt only supports 32b and 64b element types.
    unsigned LegalEltSize = EltSize <= 32 ? 32 : 64;
    if (EC == 2 || (LegalEltSize == 32 && EC == 4))
      return InstructionCost(BaseHistCntCost);

    unsigned NaturalVectorWidth = AArch64::SVEBitsPerBlock / LegalEltSize;
```

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 1 above)
Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:523]: Done.

**Review by huntergr-arm — APPROVED** [review]
LGTM, though you'll need to fix the formatting :)

**Comment by AaronBallman** [comment]
I think this caused a new build failure:
```
FAILED: lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o 
ccache /home/docker/llvm-external-buildbots/clang.17.0.6/bin/clang++ --gcc-toolchain=/gcc-toolchain/usr -DGTEST_HAS_RTTI=0 -DLLVM_EXPORTS -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/lib/Target/AArch64 -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64 -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o -MF lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o.d -o lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:21: error: logical not is only applied to the left hand side of this comparison [-Werror,-Wlogical-not-parentheses]
  537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
      |                     ^             ~~
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:21: note: add parentheses after the '!' to evaluate the comparison first
  537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
      |                     ^
      |                      (                 )
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:21: note: add parentheses around left hand side expression to silence this warning
  537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
      |                     ^
      |                     (            )
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:35: error: result of comparison of constant 32 with expression of type 'bool' is always false [-Werror,-Wtautological-constant-out-of-range-compare]
  537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
      |                     ~~~~~~~~~~~~~ ^  ~~
2 errors generated.
```
https://lab.llvm.org/buildbot/#/builders/145/builds/1925

**Comment by SamTebbs33** [comment]
> I think this caused a new build failure:
> 
> ```
> FAILED: lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o 
> ccache /home/docker/llvm-external-buildbots/clang.17.0.6/bin/clang++ --gcc-toolchain=/gcc-toolchain/usr -DGTEST_HAS_RTTI=0 -DLLVM_EXPORTS -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/lib/Target/AArch64 -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64 -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o -MF lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o.d -o lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64TargetTransformInfo.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:21: error: logical not is only applied to the left hand side of this comparison [-Werror,-Wlogical-not-parentheses]
>   537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
>       |                     ^             ~~
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:21: note: add parentheses after the '!' to evaluate the comparison first
>   537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
>       |                     ^
>       |                      (                 )
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:21: note: add parentheses around left hand side expression to silence this warning
>   537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
>       |                     ^
>       |                     (            )
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:537:35: error: result of comparison of constant 32 with expression of type 'bool' is always false [-Werror,-Wtautological-constant-out-of-range-compare]
>   537 |     if (EC == 2 || (!LegalEltSize == 32 && EC == 4))
>       |                     ~~~~~~~~~~~~~ ^  ~~
> 2 errors generated.
> ```
> 
> https://lab.llvm.org/buildbot/#/builders/145/builds/1925

Thanks Aaron, I'll commit a fix.

**Pull request merged**